### ISSUE TITLE
Allow public function to be underscored inside Controller

### ISSUE THREAD

**Issue by flashios09** [cakephp/cakephp-codesniffer#177]
Hi @markstory, 
CakePHP CS doesn't allow the public function(method) to be underscored inside Controller.
**cakephp-codesniffer/CakePHP/Sniffs/NamingConventions/ValidFunctionNameSniff.php:100**:
```php
        if ($isPublic === true) {
            if ($methodName[0] === '_') {
                $error = 'Public method name "%s" must not be prefixed with underscore';
                $phpcsFile->addError($error, $stackPtr, 'PublicWithUnderscore', $errorData);

                return;
            }
            // Underscored public methods in controller are allowed to break our rules.
            if (substr($className, -10) === 'Controller') {
                return;
            }
            // Underscored public methods in shells are allowed to break our rules.
            if (substr($className, -5) === 'Shell') {
                return;
            }
            // Underscored public methods in tasks are allowed to break our rules.
            if (substr($className, -4) === 'Task') {
                return;
            }
        }
```
The first **if** `($methodName[0] === '_')` always return the error
I didn't understand why we have the 3 other **if** !

If we want to allow the public method to be underscored inside **Controller, Shell, Task**, we have to move the first if to the end like this:
```php
        if ($isPublic === true) {
            // Underscored public methods in controller are allowed to break our rules.
            if (substr($className, -10) === 'Controller') {
                return;
            }
            // Underscored public methods in shells are allowed to break our rules.
            if (substr($className, -5) === 'Shell') {
                return;
            }
            // Underscored public methods in tasks are allowed to break our rules.
            if (substr($className, -4) === 'Task') {
                return;
            }
            if ($methodName[0] === '_') {
                $error = 'Public method name "%s" must not be prefixed with underscore';
                $phpcsFile->addError($error, $stackPtr, 'PublicWithUnderscore', $errorData);

                return;
            }
        }
```

**Comment by ADmad** [comment]
`$methodName[0] === '_'`  checks for underscore at first position. Our standards don't allow public methods starting with underscore. You can have public controller method with underscore somewhere in the middle of method name like `foo_bar`.

So you need to stop using public methods with underscore prefix if you want to confer to the CakePHP standard.

**Comment by ADmad** [comment]
While this `CakePHP.NamingConventions/ValidFunctionName` does allow public controller methods with underscore `PSR1.Methods.CamelCapsMethodName.NotCamelCaps` will still report an error for it.

But since our 2.0 standards extends PSR2 (which includes PSR1) this won't be changed. The substring checks in `ValidFunctionNameSniff` are probably leftover code for the 1.x version of code sniffer.

**Comment by flashios09** [comment]
@ADmad ok i will stick to the cakephp naming conventions but can you tell me please why we can't have an underscored public method inside controller ?

**Comment by ADmad** [comment]
> why we can't have an underscored public method inside controller

Because that's our standard :)

**Comment by flashios09** [comment]
@ADmad and i love cakephp so i have to respect **STANDARD**  ;)
